File: neural-n-tree.js
'use strict';

var NNTFunctions = NNTFunctions || {};

NNTFunctions.util = {

	assert: function(condition, message) {

        	if (! condition)
                	throw message || 'Assertation failed';

	},

	getOption: function(opt, name, def) {

        	if ('undefined' === typeof opt)
                	return def;

        	return 'undefined' !== typeof opt[name] ? opt[name] : def;

	},

	convertRange: function(value, r1, r2) {

        	var arr = [];

        	for (var i = 0, len = value.length; i < len; i++)
                	arr.push((value[i] - r1[0]) * (r2[1] - r2[0]) / (r1[1] - r1[0]) + r2[0]);

        	return arr;

	},

	similarity: function(arrOne, arrTwo) {

        	NNTFunctions.util.assert(arrOne.length === arrTwo.length, 'Array lengths do not match');

        	var a   = 0,
           	    b   = 0,
            	    c   = 0;

        	for (var i = 0, len = arrOne.length; i < len; i++) {

                	a += Math.pow(arrOne[i], 2);
                	b += Math.pow(arrTwo[i], 2);
                	c += arrOne[i] * arrTwo[i];

        	}

        	return (c / (Math.sqrt(a) * Math.sqrt(b)));

	},

	random: function(min, max) {

        	return Math.random() * (max - min) + min;

	},

	randArr: function(size, dimensions) {

        	var arr = [];

        	for (var i = 0; i < size; i++) {

                	var pointArr = [];

                	for (var j = 0; j < dimensions; j++)
                        	pointArr.push(NNTFunctions.util.random(-2, 2));

                	arr.push(pointArr);

        	}

        	return arr;

	}

};

NNTFunctions.NNTNode = function(point) {

        this.point      = point,
        this.parent     = null,
        this.left       = null,
        this.right      = null,
	this.index	= -1,
        this.isTerminal = false;

};

NNTFunctions.NeuralNTree = function(option) {

        this.rootNode   = null,
        this.alpha      = NNTFunctions.util.getOption(option, 'learningRate', 0.01),
        this.terminals  = {},
        this.index      = 0,
        this.min        = NNTFunctions.util.getOption(option, 'scaleMin', -1),
        this.max        = NNTFunctions.util.getOption(option, 'scaleMax', 1),
        this.size       = NNTFunctions.util.getOption(option, 'size', 1000),
        this.dimensions = NNTFunctions.util.getOption(option, 'dimensions', 3);

        var arr         = NNTFunctions.util.randArr(this.size, this.dimensions);

        this.initTree(arr);
        this.postOrder(this.rootNode);

};

NNTFunctions.NeuralNTree.prototype.postOrder = function(node) {

        if (null !== node && 
	    'undefined' !== typeof node) {

                this.postOrder(node.left);
                this.postOrder(node.right);

                var temp = this.index;
                node.index = temp;

                this.index++;

                if (node.isTerminal) 
                        this.terminals[temp] = node;

        }

};

NNTFunctions.NeuralNTree.prototype.initTree = function(arr) {

        return this._initTree(arr, 0, arr.length - 1);

};

NNTFunctions.NeuralNTree.prototype._initTree = function(arr, min, max) {

        if (min > max)
                return null;

        var mid = Math.round((min + max) / 2),
            node = new NNTFunctions.NNTNode(arr[mid]);

        if (null === this.rootNode)
                this.rootNode = node;

        node.left = this._initTree(arr, min, mid - 1);

        if (null !== node.left)
                node.left.parent = node;

        node.right = this._initTree(arr, mid + 1, max);

        if (null !== node.right)
                node.right.parent = node;

        if (null === node.left && 
	    null === node.right)
                node.isTerminal = true;

        return node;

};

NNTFunctions.NeuralNTree.prototype.backwardTrain = function(arg) {

        arg	= NNTFunctions.util.convertRange(arg, [this.min, this.max], [-2, 2]);

        var bmu = null;

        for (var key in this.terminals) {

                if (null === bmu)
                        bmu = this.terminals[key];

		var bmuSim = NNTFunctions.util.similarity(arg, bmu.point),
		    curSim = NNTFunctions.util.similarity(arg, this.terminals[key].point);

		if (bmuSim > curSim)
			bmu = this.terminals[key];

        }

        while (null !== bmu) {
                
		this.levelUpdate(bmu, arg);
		bmu = bmu.parent;

        }

};

NNTFunctions.NeuralNTree.prototype.levelOrder = function(node) {

	var result = [];
	
	if (null === node)
		return result;
	
	var list = [];
	
	list.push(node);
	
	while (list.length > 0) {
	
		var plist = [],
		    level = [];
			
		while (list.length > 0) {
		
			node = list.shift();
			
			level.push(node);
			
			var left  = node.left,
			    right = node.right;
				
			if (null !== left) {
			
				plist.push(left);
			
			}
			
			if (null !== right) {
			
				plist.push(right);
			
			}
		
		}
		
		result.push(level);
		
		list = plist;
	
	}

	return result;

};

NNTFunctions.NeuralNTree.prototype.levelUpdate = function(node, arg) {

        var tree 	= this.levelOrder(node),
	    levels 	= tree.length;

        for (var i = 0, len = tree.length; i < len; i++) {

                for (var j = 0, treeLen = tree[i].length; j < treeLen; j++) {
                
			if (null !== tree[i][j]) {
                                
				for (var k = 0, pointLen = tree[i][j].point.length; k < pointLen; k++) {
                                        
					tree[i][j].point[k] = this.update(tree[i][j].point[k], arg[k], ((i + 1) / (levels + 1)));
					
				}
		
			}

		}
		
        }

};

NNTFunctions.NeuralNTree.prototype.update = function(point, arg, curLevel) {

        var val = point + this.alpha * (1 / curLevel) * (arg - point);

        return val;

};

NNTFunctions.NeuralNTree.prototype.cluster = function(arg) {

        arg     = NNTFunctions.util.convertRange(arg, [this.min, this.max], [-2, 2]);

        var cur = this.rootNode;

        while (! cur.isTerminal) {

                if (null !== cur.left &&
		    null !== cur.right) {

                        var left  = NNTFunctions.util.similarity(arg, cur.left.point),
                            right = NNTFunctions.util.similarity(arg, cur.right.point);

                        if (left >= right)
                                cur = cur.left;
                        else
                                cur = cur.right;

                } else if (null === cur.left && 
			   null !== cur.right) {
                 
			cur = cur.right;
			
		} else {
			
                        cur = cur.left;
			
		}

        }

        return cur.index;

};

NNTFunctions.NeuralNTree.prototype.forwardTrain = function(arg) {

        arg	= NNTFunctions.util.convertRange(arg, [this.min, this.max], [-2, 2]);

        var cur	= this.rootNode;

        while (null !== cur) {

                if (null !== cur.left && 
		    null !== cur.right) {

                        var left  = NNTFunctions.util.similarity(arg, cur.left.point),
                            right = NNTFunctions.util.similarity(arg, cur.right.point);

                        if (left >= right) 
                                cur = cur.left;
			else 
                                cur = cur.right;
                        
			this.levelUpdate(cur, arg);

                } else if (null !== cur.left &&
			   null === cur.right) {
			
                        cur = cur.left;
			
		} else {
			
                        cur = cur.right;
			
		}

        }

};

File: neural-n-tree.py
import random
import math

class NNTNode:
	
	def __init__(self, point):
		
		self.point = point
		self.index = -1
		self.right = None
		self.left = None
		self.is_terminal = False
		self.parent = None
		
		
class NNT:

	def __init__(self, min, max, dimensions, alpha = 0.01, size = 8):
		
		self.terminals = []
		self.root = None
		self.dimensions = dimensions
		self.min = min
		self.max = max
		self.alpha = alpha
		self.index = 0
		self.size = size * 2 - 1
		
		node_array = self.__random_array()
		
		self.__init_tree(node_array)
		
		self.__post_order(self.root)
		
	
	def similarity(self, codebook, feature):
		
		a = 0.0
		b = 0.0
		c = 0.0
        
		for i in range(self.dimensions):
			a += float(codebook[i]) * float(codebook[i])
			b += float(feature[i]) * float(feature[i])
			c += float(codebook[i]) * float(feature[i])
        	
		return (c / (math.sqrt(a) * math.sqrt(b)))
		
	
	def __random(self, min, max):
		
		return random.uniform(min, max)
		
	
	def __random_array(self):
		
		node_array = []
		
		for i in range(0, self.size):
			point_array = []
			for j in range(0, self.dimensions):
				point_array.append(self.__random(self.min, self.max))
			node_array.append(point_array)
		
		return node_array
		
	
	def __init_tree(self, point_array):
		
		return self.__init_tree_recursive(0, len(point_array) - 1, point_array)
		
	
	def __init_tree_recursive(self, min, max, point_array):
		
		if min > max:
			return None

		mid = int(round((min + max) / 2))
		node = NNTNode(point_array[mid])
		
		if self.root == None:
			self.root = node
			
		node.left = self.__init_tree_recursive(min, mid - 1, point_array)
			
		if node.left != None:
			node.left.parent = node
        		
		node.right = self.__init_tree_recursive(mid + 1, max, point_array)
        
		if node.right != None:
			node.right.parent = node
        	
		if node.right == None and node.left == None:
			node.is_terminal = True
        	
		return node
        
		
	def level_order(self, node):
	
		temporary = []
		tree = []
		nodes = []
		current = node
		level = 1
		
		temporary.append(current)
		tree.append([current])
		
		while len(temporary) > 0:
		
			if len(nodes) >= 2 ** level:
			
				tree.append(nodes)
				nodes = []
				level += 1	
			
			current = temporary.pop(0)
			
			if current != None:
			
				left = current.left
				right = current.right
				
				temporary.append(left)
				nodes.append(left)
				
				temporary.append(right)
				nodes.append(right)
				
			else:
			
				nodes.append(None)
				nodes.append(None)
				
		if len(nodes) > 0:
			tree.append(nodes)
			
		return tree
		
		
		
	def __post_order(self, node):
		
		if node != None:
		
			self.__post_order(node.left)
			self.__post_order(node.right)
			
			index = self.index
			node.index = index
			
			self.index += 1
			
			if node.is_terminal:
				self.terminals.append(node)
				
				
		
	def forward_train(self, arg):
		
		current = self.root
		
		self.__level_update(current, arg)
		
		while not current.is_terminal:
		
			if current.left != None and current.right != None:
			
				left = self.similarity(current.left.point, arg)
				right = self.similarity(current.right.point, arg)
				
				if left >= right:
				
					current = current.left
					
				else:
				
					current = current.right
					
				self.__level_update(current, arg)
					
			elif current.left != None and current.right == None:
			
				curret = current.left
			
			else:
				
				current = current.right
		
		
	def backward_train(self, arg):
		
		bmu = None

		for node in self.terminals:
			
			if bmu == None:
				bmu = node
		
			bmu_similarity = self.similarity(arg, bmu.point)
			node_similarity = self.similarity(arg, node.point)
		
			if bmu_similarity > node_similarity:
				bmu = node

		while bmu != None:
        	
			self.__level_update(bmu, arg)
			bmu = bmu.parent

        
                

		
	def cluster(self, arg):
		
		current = self.root
		
		while not current.is_terminal:
		
			if current.left != None and current.right != None:
			
				left = self.similarity(current.left.point, arg)
				right = self.similarity(current.right.point, arg)
				
				if left >= right:
				
					current = current.left
					
				else:
				
					current = current.right
					
			elif current.left != None and current.right == None:
			
				curret = current.left
			
			else:
				
				current = current.right
				
		return current.index
		
		
		
	def __level_update(self, node, arg):
	
		tree = self.level_order(node)
		levels = len(tree)
		
		
		for i in range(levels):
			for j in range(len(tree[i])):
				if tree[i][j] != None:
					for k in range(len(tree[i][j].point)):
						tree[i][j].point[k] = self.__update(tree[i][j].point[k], arg[k], (float((i + 1)) / (float(levels + 1))))
						
						
		
	def __update(self, point, arg, level):
		
		return point + self.alpha * (1 / level) * (arg - point)
		
